_drafts/building-a-cli-for-firmware-projects.md

```diff
@@ -0,0 +1,515 @@
+script, or continuous integration system to perform the most common operations
+are performed everyday.
+
+For example, it helps convert launching JLinkGDBServer with the appropriate
+settings from a Wiki copy-paste of:
+
+```
+$ JLinkGDBServer -device Cortex-M4 -if SWD -speed 4000 -RTOSPlugin_FreeRTOS
+```
+
+to
+
+```
+invoke jlink
+```
+
+and flashing using GDB from:
+
+```
+$ cd products/blinky/gcc
+$ arm-none-eabi-gdb-py -x gdbinit.jlink ../build/product.elf -x ../../common_gdb_scripts.gdbinit
+```
+
+to
+
+```
+invoke flash
+```
+
+A web service's REST API needs to be stable, easy to use, and self documenting.
+Your project's CLI should meet the same requirements.
+
+## When to Build Project CLI
+
+If you or your teammates have experienced any of the following more than once,
+it may be time to write a CLI for your project.
+
+- You **copy/paste** commands from a Wiki page or an old Slack message
+- You found the command you were looking for but it's now **out-of-date**
+- It has become common for teammates to ask in Slack **"What is that
+  command..."**
+- Scripts have **undocumented** arguments and environment variable overrides
+- Two or more teammates have **written similar scripts** for themselves but a
+  common shared version doesn't exist, or is broken.
+- You check out a 3 month-old revision of your project and there is no longer
+  documentation about how to run (factory builds anyone?).
+
+I could go on and on, but I hope the idea is clear. Building a CLI using Invoke
+helps with all of the above issues and more.
+
+## Getting started with Invoke
+
+Invoke is a "task execution tool & library" that works with Python 2.7 and 3.4+,
+and be installed using `pip`. We'll be using it today to build our project's CLI
+tool.
+
+We'll be using a virtual environment, and I highly recommend them.
+[This guide](https://docs.python-guide.org/dev/virtualenvs/#lower-level-virtualenv)
+is a good starting point. I will be using Python 3.6, which has the `virtualenv`
+command available already.
+
+```
+$ virtualenv venv
+$ source venv/bin/activate
+(venv)
+$ pip install invoke
+```
+
+We can quickly test if it was properly installed by running `invoke` in our
+shell.
+
+```
+(venv)
+$ invoke --list
+Can't find any collection named 'tasks'!
+```
+
+We've confirmed it's installed. To start adding tasks to Invoke, one needs to
+add a `tasks.py` file to the root of a project. No matter where the user is
+within the project, Invoke will search upwards until it finds a `tasks.py` file.
+
+## Using Invoke with the nRF5 SDK
+
+> For purposes of this example, we are going to use the Nordic nRF5 SDK version
+> 15.2.0, which can be downloaded
+> [here](https://developer.nordicsemi.com/nRF5_SDK/nRF5_SDK_v15.x.x/), and
+> within that, the Blinky app located at
+> `nrf5_sdk/examples/peripheral/blinky/pca10056/blank/armgcc/`.
+
+The following is the most basic, but still reasonable, `tasks.py` to allow us to
```

Comment:
I'd remove "but still reasonable". 

-> The following is a simple `tasks.py` we have written to make it easier to buld our Blinky project